Doc/whatsnew/3.10.rst

```diff
@@ -135,17 +135,35 @@ Porting to Python 3.10
 
 * Since :c:func:`Py_TYPE()` is changed to the inline static function,
   ``Py_TYPE(obj) = new_type`` must be replaced with ``Py_SET_TYPE(obj, new_type)``:
-  see :c:func:`Py_SET_TYPE()` (available since Python 3.9).
+  see :c:func:`Py_SET_TYPE()` (available since Python 3.9). For backward
+  compatibility, this macro can be used::
+
+      #if PY_VERSION_HEX < 0x030900A4
+      #  define Py_SET_TYPE(obj, type) do { Py_TYPE(obj) = (type); } while (0)
+      #endif
```

Comment:
This macro doesn't behave the same way as the function in 3.9. I think a more accurate macro would be:

```suggestion
      #if PY_VERSION_HEX < 0x030900A4
      #  define Py_SET_TYPE(obj, type) ((Py_TYPE(obj)) = (type), (void)0)
      #endif
```

which can be used as part of an expression like the function version of `Py_SET_TYPE` can.

Reply:
Thanks for finding some precedent. I've seen "always use parens and `do { } while(0);`" given as macro advice forever, but I don't remember ever seeing this case before.

Reply:
In the meanwhile, I created the backport compatiblity header file :-) It defines a static inline function which returns void:
https://github.com/pythoncapi/pythoncapi_compat/blob/219cdd858595fb5d478117b0ef4758fedc3d2365/pythoncapi_compat.h#L29

Maybe tomorrow when it will be more mature (than 1 hour old), numpy might want to use it.